solr/core/src/java/org/apache/solr/cluster/placement/ReplicaPlacement.java

```diff
@@ -0,0 +1,29 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.solr.cluster.placement;
+
+/**
+ * <p>Placement decision for a single {@link Replica}. Note this placement decision is used as part of a {@link WorkOrder},
+ * it does not directly lead to the plugin code getting a corresponding {@link Replica} instance, nor does it require the
+ * plugin to provide a {@link Shard} instance (the plugin code gets such instances for existing replicas and shards in the
+ * cluster but does not create them directly for adding new replicas for new or existing shards).
+ *
+ * <p>Captures the {@link Shard} (via the shard name), {@link Node} and {@link Replica.ReplicaType} of a Replica to be created.
+ */
+public interface ReplicaPlacement {
```

Comment:
I think we must be more explicit here - the placement refers to Request (very important for debugging and audit of the plugin decisions!), and it should at the very least include the target node.

Reply:
It does include the `Node`. See `WorkOrderFactory.createReplicaPlacement()`. It does not directly refer to a `Request`, the reference to `Request` is captured in the `WorkOrder` created using the same factory and in which the `ReplicaPlacement` are used.
Everything passed to the creation factories can be made accessible on the returned instances if needed (given it's captured in the underlying implementations), but I'm not convinced it's useful, so kept it simple. Assumption is that plugin code creates these instances so plugin code knows why and keeps track of what each created instance refers to... But again, easy to add here and in other instances returned by factories (might need to define subinterfaces then to make the appropriate values accessible - BTW that's how I started coding locally, but then simplified to limit the number of interfaces and for the reasons exposed above).

Reply:
It's a common problem today (or at least in 8x) to actually understand why the policy made the placement decisions it did. Imagine having to investigate it without having a handy reference to the request that caused these decisions.

So the purpose of this reference is not for the plugin, it's for the user to be able to track why a particular placement decision was made. We could leave this to the plugins to implement it but I think it's an important aspect of auditing and debugging and we should enforce it on plugins using the interfaces.

Even if a plugin just randomly assigns placements there is always a cause for the assignments to be generated. Whether we include the whole Request or just an id, I think we should explcitly require this information to be tracked.

Reply:
It is already tracked. As the plugin creates and returns a `WorkOrder` to tell Solr where to do what, it passes a reference to the `Request`. Solr code would be totally able to log the request and the corresponding placement decisions. `ReplicaPlacement` instances do not exist outside a `WorkOrder` in Solr (they do in plugin code but if not added to a `WorkOrder` it's as if they were not created).

Reply:
> WorkOrder to tell Solr where to do what, it passes a reference to the Request
Hmm, WorkOrder is an empty interface. I think we should add it there explicitly, either the Request or at least the id.

Reply:
Ok, added. Note that the interfaces in `org.apache.solr.cluster.placement` only condition what the plugin code is seeing, not what's present in the objects. The `Request` is present in the `WorkOrder` instances and I added it to the interface, but even if it's not in the interface, Solr side code would have it.
My thinking (by not putting it in the interface) is that plugin code is creating these instances so it should know which instance is which and doesn't need a method on that instance to find out.

Reply:
Right, the plugin doesn't need this - but other parts of Solr may need it to simplify error handling if the error is too far away in the code from the original Request.

Reply:
Solr code will not be using these interfaces. This will be clearer once I'm back from vacation and implement the solr side of things. Solr will only be using the concrete implementations of the interfaces so will have access to everything it needs.